### ISSUE TITLE
Images are not properly registered

### ISSUE THREAD

**Issue by znnnnn** [DiffuseMorph/DiffuseMorph#12]
Hi,
Does anyone use the author's toy_sample data for testing like me, and the result is that the generated moved image is exactly the same as the original moving image, and it seems that the deformation field is not generated correctly.
And I used the complete RAFD data set, trained and tested according to the description in the paper, and still reproduced this problem.

**Comment by Barrett-python** [comment]
是的 我也是这个问题，我训练完用测试图片推理，没有出现deformation field。请问你解决了吗

**Comment by Atoposiii** [comment]
> Hi, Does anyone use the author's toy_sample data for testing like me, and the result is that the generated moved image is exactly the same as the original moving image, and it seems that the deformation field is not generated correctly. And I used the complete RAFD data set, trained and tested according to the description in the paper, and still reproduced this problem.

can you help me？

**Comment by ZhaiJiaKai** [comment]
Yes, I'm running into this issue as well, do you guys have a solution？

**Comment by evarcurto** [comment]
Has anyone here been able to successfully train this network? Every time I give it a go, the output ends up being a solid yellow square. Any tips or tricks you've discovered?

**Comment by ZhaiJiaKai** [comment]
Yes, I'm having this issue as well, have you solved it now, please

**Comment by loladidi** [comment]
Are you operating on a Linux system or a Windows system?

**Comment by loladidi** [comment]
> 是的，我也有这个问题，你现在解决了吗，请

Are you operating on a Linux system or a Windows system?

**Comment by evarcurto** [comment]
Linux

**Comment by loladidi** [comment]
> 嗨，有没有人像我一样使用作者的toy_sample数据进行测试，结果是生成的移动图像与原始移动图像完全一样，并且似乎没有正确生成变形场。我使用了完整的RAFD数据集，根据论文中的描述进行了训练和测试，仍然重现了这个问题。

![image](https://github.com/DiffuseMorph/DiffuseMorph/assets/162962299/c092a4dd-a65c-408c-a67e-7664a40a6803)

Hello, I also received this result, but it was not obtained through testing, but after training. My test did not run successfully, but encountered the following issue

![image](https://github.com/DiffuseMorph/DiffuseMorph/assets/162962299/9f570b2c-5553-4cd6-936a-21726114cfd4)

Are you still doing this research? Can you exchange ideas

**Comment by loladidi** [comment]
> 这里有人能够成功训练这个网络吗？每次我试一试，输出最终都会变成一个实心黄色方块。您发现了任何提示或技巧吗？

May I ask if your testing process was successful？

**Comment by evarcurto** [comment]
The error you get in the test is because the trained model file doesn't exist. Training the toy sample gives me the same thing. I've been trying to train models with the full dataset (RaFD dataset), trying to recreate the training conditions mentioned in the article. But so far, no success. 
![Screenshot from 2024-03-13 11-44-49](https://github.com/DiffuseMorph/DiffuseMorph/assets/30838057/15198071-0d0f-40ee-8df8-6f085956e484)
![Screenshot from 2024-03-13 11-45-10](https://github.com/DiffuseMorph/DiffuseMorph/assets/30838057/2a17149c-f83e-4ae5-937d-c4c0ab19984c)
![Screenshot from 2024-03-13 11-45-31](https://github.com/DiffuseMorph/DiffuseMorph/assets/30838057/437eb7c5-3f35-45cb-a928-de1c76b7cd5d)
![Screenshot from 2024-03-13 11-45-37](https://github.com/DiffuseMorph/DiffuseMorph/assets/30838057/ce59ebce-b767-47de-b515-25ff7b290007)

**Comment by loladidi** [comment]
> The error you get in the test is because the trained model file doesn't exist. Training the toy sample gives me the same thing. I've been trying to train models with the full dataset (RaFD dataset), trying to recreate the training conditions mentioned in the article. But so far, no success. ![Screenshot from 2024-03-13 11-44-49](https://private-user-images.githubusercontent.com/30838057/312439366-15198071-0d0f-40ee-8df8-6f085956e484.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzMzE2MTgsIm5iZiI6MTcxMDMzMTMxOCwicGF0aCI6Ii8zMDgzODA1Ny8zMTI0MzkzNjYtMTUxOTgwNzEtMGQwZi00MGVlLThkZjgtNmYwODU5NTZlNDg0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzEzVDEyMDE1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM4YWYwMzc4YzQ4NzRiYWE1YTY3ZWZhOGU5NTI0ODYzZTA2MTA0ODcxNDM2NzUwZWEyZGY1NGZhN2MyNTlkOWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.q79oe0Joj15yIkiw2UieJYfFU3TvXfdQyhQfYfgpxiQ) ![Screenshot from 2024-03-13 11-45-10](https://private-user-images.githubusercontent.com/30838057/312439360-2a17149c-f83e-4ae5-937d-c4c0ab19984c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzMzE2MTgsIm5iZiI6MTcxMDMzMTMxOCwicGF0aCI6Ii8zMDgzODA1Ny8zMTI0MzkzNjAtMmExNzE0OWMtZjgzZS00YWU1LTkzN2QtYzRjMGFiMTk5ODRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzEzVDEyMDE1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJkYWI0YjJiYjRhZDlhYmU2OWJkNjk1NjdjYzNkYWEzYTdlNGY3ZDQ1MmEyNWM0OGE3Yjk0Y2RlNjdkOGQzODYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OwByujtcWLZkJtmyTc-GEJHsyMrIw4p3d3BsKw8NxyU) ![Screenshot from 2024-03-13 11-45-31](https://private-user-images.githubusercontent.com/30838057/312439341-437eb7c5-3f35-45cb-a928-de1c76b7cd5d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzMzE2MTgsIm5iZiI6MTcxMDMzMTMxOCwicGF0aCI6Ii8zMDgzODA1Ny8zMTI0MzkzNDEtNDM3ZWI3YzUtM2YzNS00NWNiLWE5MjgtZGUxYzc2YjdjZDVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzEzVDEyMDE1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwZTAzZWZiOTc3ZGNmYTE5M2YxMWRkODEzMDc1ZThlNTJhNzBjY2NjMjVkMWZiMjNmZWU2YzIwYTA0NDZjMjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.p5WAxHaHAlqn9nQOMG13i-dZ5Vul7J3m-UPihQopwik) ![Screenshot from 2024-03-13 11-45-37](https://private-user-images.githubusercontent.com/30838057/312439373-ce59ebce-b767-47de-b515-25ff7b290007.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzMzE2MTgsIm5iZiI6MTcxMDMzMTMxOCwicGF0aCI6Ii8zMDgzODA1Ny8zMTI0MzkzNzMtY2U1OWViY2UtYjc2Ny00N2RlLWI1MTUtMjVmZjdiMjkwMDA3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzEzVDEyMDE1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU2OTU0ZDk2N2JkZjdmMDNmNzNhNGJlMjVhZWUxNjg1Y2RkZmFiOGQxMjc4MDdlYjJkMjQxODI5NDA4ZGZkOTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.qGTlXbT762nRlMh-w3zNwAJmJyS5PNVZuI4xO7nW0jY)

Thank you very much for your answer. The fact is indeed so. After correction, I encountered the same problem as everyone else: it seems that the program has not been registered.
![image](https://github.com/DiffuseMorph/DiffuseMorph/assets/162962299/7bd64ae9-de07-4960-88dd-6b0e6f7c639a)
Additionally, I noticed that there is another work on GitHub that he has made some changes, which may be helpful for your research.
![image](https://github.com/DiffuseMorph/DiffuseMorph/assets/162962299/1d1b43ee-b1f3-409e-9352-c150062f2193)

**Comment by evarcurto** [comment]
Do you think it makes sense to use batch size = 1?

**Comment by loladidi** [comment]
> 您认为使用批量大小= 1 有意义吗？

What does it mean? Is there any problem? Where is the batch size=1?

**Comment by evarcurto** [comment]
They don't mention in the paper what batch size is used in the training sessions, but in the original .json files the batch size is set to 1. I don't have much experience in Deep Learning (i.e. tuning hyperparameters). That's why I question whether setting the batch size to 1 makes sense, given that it takes many days to run a model.
I've asked the authors these questions by email but they haven't replied.

**Comment by loladidi** [comment]
> They don't mention in the paper what batch size is used in the training sessions, but in the original .json files the batch size is set to 1. I don't have much experience in Deep Learning (i.e. tuning hyperparameters). That's why I question whether setting the batch size to 1 makes sense, given that it takes many days to run a model. I've asked the authors these questions by email but they haven't replied.

I took a look at the repository again and found that the author manually increased the batch size for generalization. You can take a look at his readme. In fact, he tried training a total of 560/40=14 pairs of images.
In addition, I would like to ask if you can now generate an accurate deformation field for a set of images (instead of generating a yellow square) and register it to achieve a similar effect as in the article, because I have not been able to generate the correct deformation field yet. I am not sure why it is a yellow square. Is there any problem during the training process, or is there an incorrect parameter setting. If you have already implemented it, can you give me an answer? After all, I haven't even succeeded with a set of images, so I don't have the time to consider generalization issues.

**Comment by loladidi** [comment]
> They don't mention in the paper what batch size is used in the training sessions, but in the original .json files the batch size is set to 1. I don't have much experience in Deep Learning (i.e. tuning hyperparameters). That's why I question whether setting the batch size to 1 makes sense, given that it takes many days to run a model. I've asked the authors these questions by email but they haven't replied.

If everyone is unable to generate a deformation field, perhaps this method is not feasible.

**Comment by evarcurto** [comment]
To be honest I don't understand the training methodology of sinanseslikaya. So far, I can get images like this 
![image](https://github.com/DiffuseMorph/DiffuseMorph/assets/30838057/6d3b2a97-7634-427a-a534-a4fd2a92a769)
@loladidi What do you think about scheduling a call to share ideas more fluidly?

**Comment by loladidi** [comment]
> To be honest I don't understand the training methodology of sinanseslikaya. So far, I can get images like this ![image](https://private-user-images.githubusercontent.com/30838057/313173160-6d3b2a97-7634-427a-a534-a4fd2a92a769.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA1MDU1ODksIm5iZiI6MTcxMDUwNTI4OSwicGF0aCI6Ii8zMDgzODA1Ny8zMTMxNzMxNjAtNmQzYjJhOTctNzYzNC00MjdhLWE1MzQtYTRmZDJhOTJhNzY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE1VDEyMjEyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEwNjY1YzExOWNmMTlmZTY3Y2JjNmRlN2NlNWJlZmU0YzA5Y2U5MWZmYzQ5MTJkZGIxN2I5MzA2ZDM3ZmEyZmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.lRo-y-qDNqGywbnU2SIJ8Eyl22s52n4UeKNJjBFhHCU) @loladidi What do you think about scheduling a call to share ideas more fluidly?

I see that you have received a very good epoch001_flow. png. Can I take a look at the effect image you have obtained epoch001_out-M. png?
In fact, Sinaneslikaya's training method is not good, it is just a reference.
I just started looking at this method, and now I may not be able to communicate much through phone calls. You can wait for me to study it for a while.

**Comment by evarcurto** [comment]
![Screenshot from 2024-03-15 12-56-06](https://github.com/DiffuseMorph/DiffuseMorph/assets/30838057/5055c31a-f1cc-4692-ada6-a39ca26604dc)
Ok, take your time. I already made some modifications to the original code like a new function to read pairs of images from several subfolders (subfolders regarding the whole RafD dataset). I think to reproduce the results of the article it only makes sense to train with the complete dataset.
My code is messy but if I have time to clean it I can later share it with you.

**Comment by loladidi** [comment]
> ![Screenshot from 2024-03-15 12-56-06](https://private-user-images.githubusercontent.com/30838057/313199963-5055c31a-f1cc-4692-ada6-a39ca26604dc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA1MTA4ODYsIm5iZiI6MTcxMDUxMDU4NiwicGF0aCI6Ii8zMDgzODA1Ny8zMTMxOTk5NjMtNTA1NWMzMWEtZjFjYy00NjkyLWFkYTYtYTM5Y2EyNjYwNGRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE1VDEzNDk0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY2ZmY2ZDBkOGIwNWQyM2JkM2IwYzZiMTVkMjVhZmJlYTc0YmYzZTgwNTZjYTE4OWI4MjViOGY1YTJhMTBkY2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.tBYmd4cDF23AXet9rOSggccm_j3UrAjX6AidiWu6GQM) Ok, take your time. I already made some modifications to the original code like a new function to read pairs of images from several subfolders (subfolders regarding the whole RafD dataset). I think to reproduce the results of the article it only makes sense to train with the complete dataset. My code is messy but if I have time to clean it I can later share it with you.

I fully agree with your point of view. From your results, it seems that you have indeed done a great job. If you are willing to share your code, it would be very helpful for my work, and I would greatly appreciate it. Looking forward to your sharing and future communication.

**Comment by evarcurto** [comment]
My attempt to use DiffuseMorph: https://github.com/evarcurto/myDiffuseMorph 
- Class RAFDDataset modified to obtain image pairs
- Schedule error fixed
- Modifications to run with batch size > 2

**Comment by loladidi** [comment]
> 我尝试使用 DiffuseMorph：https://github.com/evarcurto/myDiffuseMorph
> 
> * 修改了 RAFDDataset 类以获取图像对
> * 计划错误已修复
> * 修改为以批处理大小 > 2 运行

Hello, thank you very much for sharing, but I still haven't generated the correct results. I'm not sure if I still haven't operated correctly. I also noticed that in your last picture, the moved image still seems to be the same as the moving image. Although there is a deformation field, the moved image is not registered with the fixed image.
![836ed501959e3f36668987fb22e9668](https://github.com/DiffuseMorph/DiffuseMorph/assets/162962299/a716006a-3de5-40f9-98f1-1a7082ff9862)

**Comment by evarcurto** [comment]
I said that my attempts to train the model were unsuccessful (https://github.com/DiffuseMorph/DiffuseMorph/issues/12#issuecomment-1994195016 ). The best I get from the flow images are "ghosts" similar to moving images. Are you already training with the RaFD dataset (with the cropped and resized images)? It's normal for the first epoch not to be able to generate the correct flow. That's why the model has to learn for several epochs. But it's expected the model can get the correct flow by the 40 epoch, which is not happening in my attempts. Sometimes the errors stay with Nan and I have to abort the training (see the graph in the last printscreen), other times I get to the end without being able to estimate the flow (see the next printscreen). ![Screenshot from 2024-03-08 13-50-31](https://github.com/DiffuseMorph/DiffuseMorph/assets/30838057/1f46a5e1-3f32-4211-a15d-b5de910a77fb)
